diff --git a/GUI.cpp b/GUI.cpp
index 6af2312..650f5e6 100644
--- a/GUI.cpp
+++ b/GUI.cpp
@@ -658,7 +658,7 @@ appEventResult GUI::processEvent(uint32_t now, EventType event) {
             state.locked = false;
             res |= REDRAW_SCREEN | LOCK_UNLOCK;
           }
-          if (state.unlockButton1) {
+          if (state.unlockButton1 && event != WIPHONE_KEY_UP & event != WIPHONE_KEY_DOWN && event != WIPHONE_KEY_LEFT && event != WIPHONE_KEY_RIGHT ) {
             state.unlockButton1 = 0;
             res |= REDRAW_FOOTER;
             log_d("state.unlockButton1 cleared");

