diff --git a/lora.cpp b/lora.cpp
index 6b03536..ae1833b 100644
--- a/lora.cpp
+++ b/lora.cpp
@@ -43,7 +43,20 @@ void Lora::setup() {
   loraSPI->setPins(HSPI_MISO, HSPI_MOSI, HSPI_SCLK);
   pinMode(RFM95_RST, OUTPUT);
   rf95->init();
-  rf95->setFrequency(RF95_FREQ);
+
+
+  float freq = RF95_FREQ;
+  // Load phone configs
+  {
+    CriticalFile ini(Storage::ConfigsFile);
+    if ((ini.load() || ini.restore()) && !ini.isEmpty()) {
+      if (ini[0].hasKey("v") && !strcmp(ini[0]["v"], "1")) {    // check version of the file format
+        if ( ini.hasSection("lora") )
+          freq = ini["lora"].getFloatValueSafe("lora_freq", RF95_FREQ);
+      }
+    }
+  }
+  rf95->setFrequency(freq);
   rf95->setTxPower(23, false);
 
   log_v("Free memory after LoRa: %d %d", ESP.getFreeHeap(), heap_caps_get_free_size(MALLOC_CAP_32BIT));
